### PR TITLE
Tests for module

### PR DESCRIPTION
**Pull request by colincoleman**

### CODE REVIEW

File: tests/default/test.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+set -eo pipefail
+tests_failed=0
+cd terraform
+terraform init
+terraform apply --auto-approve
+VPC_ID=`terraform output -json | jq -r '.vpc_id.value'`
+subnet_count=`aws ec2 describe-subnets | jq --arg VPC_ID "$VPC_ID" '.Subnets[]| select (.VpcId==$VPC_ID)' | jq -s length`
+terraform destroy --auto-approve
+
+if [ $subnet_count -eq 3 ]
```
Comment (itsdalmo) [inline tests/default/test.sh]: We don't default to 1x public and private subnet per AZ anymore?

File: tests/private-subnets-no-nat/test.sh [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+if [ $natgateway_count -eq 0 ]
+ then
+   echo "√ Expected # of Nat Gateways"
+ else
+   echo "✗ Expected # of Nat Gateways"
+   tests_failed=$((tests_failed+1))
+fi
+
+if [ $egress_only_igw_count -eq 1 ]
```
Comment (itsdalmo) [inline tests/private-subnets-no-nat/test.sh]: Perhaps the checks could be made into functions to avoid repetition?

File: tests/default/test.sh [hunk 3]
```diff
@@ -1,5 +1,18 @@
 #!/bin/sh
 set -eo pipefail
+
+# for integers comparisons: checkCounts <testValue> <expectedValue> <testName>
+checkCounts() {
```
Comment (itsdalmo) [inline tests/default/test.sh]: I think (not 100%) that snake-case is preferred over camel-case for bash scripts 👍

File: tests/default/test.sh (hunk 1 above)
Comment (colincoleman) [reply tests/default/test.sh]: I'll leave improving the module to future PRs

File: tests/private-subnets-no-nat/test.sh (hunk 2 above)
Comment (colincoleman) [reply tests/private-subnets-no-nat/test.sh]: Fixed

File: tests/default/test.sh (hunk 3 above)
Comment (colincoleman) [reply tests/default/test.sh]: You could be right....  I have Java muscle memory that makes me write functions that way.... I'll check and update if appropriate

Comment (colincoleman) [reply tests/default/test.sh]: You are right (according to google)
https://google.github.io/styleguide/shell.xml#Function_Names

**Comment by itsdalmo** [comment]
I think in these cases, it would be a tad bit nicer to extend the `Makefile` so we had a make target which loops over the examples and runs `terraform apply`, your tests and then `terraform destroy`.

Main benefits being:
- Easy to run locally.
- We can just drop more examples in there and not have to update the pipeline.

**Comment by colincoleman** [comment]
I can see the benefits of adding to the makefile - but there are also a few downsides:
A special docker image for testing would need to be made / maintained (terraform and awscli in same container)
You would be locked to only doing what that one (monster testing) docker file can do in any test

**Comment by itsdalmo** [comment]
> A special docker image for testing would need to be made / maintained (terraform and awscli in same container)

Good point! With that in mind I agree that this is the best solution 🙌

**Pull request merged**